### PR TITLE
[KT-15] Unsubscribe inactive People

### PR DESCRIPTION
**Pull request by nugriz**
add script that ready to be run to unsubscribe inactive people to reduce cost of sending email

### CODE REVIEW

**Review by donjar — COMMENTED** [review]
Can you also make the script send an email to the user, basically saying "We noticed that you have been inactive so we are unsubscribing you" something like that?

File: scripts/unsubscribe_inactive_people.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+
+def unsubscribe
+  User.all.each do |user|
+    next unless user.user_contests.order(:created_at).last.created_at < Time.current - 6.months
+
+    user.user_notifications.each(&:destroy!)
+    Mailgun.send_message(
+      to: user.email,
+      text: 'With this email, we apologize for your inconvenience.' \
```
Comment (donjar) [inline scripts/unsubscribe_inactive_people.rb]: Please use Bahasa Indonesia

File: test/scripts/unsubscribe_inactive_people_test.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    assert_equal UserNotification.all.size, 1
+    assert_equal UserNotification.first.user_id, @user.id
+    assert_includes UserNotification.all, @un
+    assert_raises ActiveRecord::RecordNotFound do
+      UserNotification.find(@un2.id)
+    end
+  end
+
+  private def create_items
+    load './scripts/unsubscribe_inactive_people.rb'
```
Comment (donjar) [inline test/scripts/unsubscribe_inactive_people_test.rb:20]: If this is to be inside this function, I suggest renaming the `create_items` function as it is no longer about just creating items; it also loads the script.

File: scripts/unsubscribe_inactive_people.rb [hunk 3]
```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
+
+def unsubscribe
+  User.all.each do |user|
+    next unless user.user_contests.order(:created_at).last.created_at < Time.current - 6.months
+
+    user.user_notifications.each(&:destroy!)
+    Mailgun.send_message(
+      to: user.email,
+      text: 'Dengan email ini, kami mohon maaf atas ketidaknyamanannya.' \
```
Comment (donjar) [inline scripts/unsubscribe_inactive_people.rb]: Please check this text

Comment (donjar) [reply scripts/unsubscribe_inactive_people.rb]: Kami melihat bahwa Anda sudah tidak mengikuti KTOM selama 6 bulan terakhir. Oleh karena itu, kami akan stop memberikan Anda notifikasi seputar kontes di KTOM.

Jika Anda ingin kembali subscribe ke notifikasi KTOM, Anda bisa masuk ke (link). Terima kasih.

File: scripts/unsubscribe_inactive_people.rb [hunk 4]
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
+def unsubscribe
+  User.all.each do |user|
+    next unless user.user_contests.order(:created_at).last.created_at < Time.current - 6.months
+
+    user.user_notifications.each(&:destroy!)
+    Mailgun.send_message(
+      to: user.email,
+      text: 'Kami melihat bahwa Anda sudah tidak mengikuti KTOM selama 6 bulan terakhir.' \
```
Comment (donjar) [inline scripts/unsubscribe_inactive_people.rb]: Can you put the spaces at the back? So that it is consistent with other multiline strings.

Also, change the word "stop" to "berhenti".

Otherwise, this looks good!

File: scripts/unsubscribe_inactive_people.rb [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+  User.all.each do |user|
+    next unless user.user_contests.order(:created_at).last.created_at < Time.current - 6.months
+
+    user.user_notifications.each(&:destroy!)
+    Mailgun.send_message(
+      to: user.email,
+      text: 'Kami melihat bahwa Anda sudah tidak mengikuti KTOM selama 6 bulan terakhir. ' \
+            "Oleh karena itu, kami akan berhenti memberikan Anda notifikasi seputar kontes di KTOM.\n\n" \
+            'Jika Anda ingin kembali subscribe ke notifikasi KTOM, Anda bisa masuk ke ' \
+            "https://ktom.tomi.or.id/users/#{user.id}/user-notifications. Terima kasih."
```
Comment (donjar) [inline scripts/unsubscribe_inactive_people.rb]: Use the relevant route link, i.e. `user_notifications_path` or something. (It's the one you can see by running `rake routes`.)

Comment (nugriz) [reply scripts/unsubscribe_inactive_people.rb]: I can't use _path neither _url in script, rails pop up undefined method for it, can u help me, please? @donjar

File: test/controllers/user_notifications_controller_test.rb [hunk 6]
```diff
@@ -6,6 +6,8 @@ class UserNotificationsControllerTest < ActionController::TestCase
   setup :login_and_be_admin, :create_items
 
   test 'routes' do
+    assert_equal user_user_notifications_url(@user),
```
Comment (donjar) [inline test/controllers/user_notifications_controller_test.rb]: Why is this added?

Comment (nugriz) [reply test/controllers/user_notifications_controller_test.rb]: cuz i wan't to prove that user_user_notifications_url(@user) is the right link instead of user_user_notification_path(@user). I can remove it if you @donjar want

Comment (donjar) [reply test/controllers/user_notifications_controller_test.rb]: I understand your intentions, and I think it's good that you have an eye for testing. Nevertheless here are several reasons why I think this is not a good idea -
1. You included `test.host` here which might change in different environments. If this host name somehow changes (for example, maybe upgrading Rails somehow changes it), this test will break.
2. We have tested `user_user_notifications_path` below, which is essentially the same.
3. The exact URL, which you are testing here, is actually not relevant to the logic of the unsubscribe function. We don't necessarily care which exact link the unsubscribe link is pointing to - we only care that it points to the one under `user_user_notifications_url`.

Given these considerations I suggest you remove this test.

Comment (nugriz) [reply test/controllers/user_notifications_controller_test.rb]: I got it, thanks for the enlightenment

File: scripts/unsubscribe_inactive_people.rb [hunk 7]
```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+include Rails.application.routes.url_helpers
```
Comment (donjar) [inline scripts/unsubscribe_inactive_people.rb]: Can you try moving this into the `unsubscribe` method? See if that works

Comment (nugriz) [reply scripts/unsubscribe_inactive_people.rb]: if u @donjar see one commit before this, which is "fix undefined path", i put it inside ```unsubscribe``` method but the error is undefined method include. so i put it outside and found this error, i guess

**Review by donjar — APPROVED** [review]
